### ISSUE TITLE
The data types varchar and nvarchar are incompatible in the '^' operator

### ISSUE THREAD

**Issue by miqueljuan** [dotnet/efcore#35093]
Projection to Boolean of Equals comparator of a Enum property (that has custom ValueConverter to string) with one of possibles values of Enum translates to SQL with the new XOR ('^') operator implementation, causing 'The data types varchar and nvarchar are incompatible in the '^' operator'.

```C#
dbContext.Orders.Select(order=> new OrderModel {  Paid = order.State == OrderState.Paid }).ToList();

enum OrderState { Unpaid, Paid, ... }

class OrderStateConverter : ValueConverter<OrderState, string> 
{
  private static string ConvertToString(OrderState state)
  {
      return stateswitch
      {
          OrderState.Unpaid=> "U",
          OrderState.Paid => "P",
          ....
          _ => throw new InvalidEnumArgumentException(nameof(state), (int)state, typeof(OrderState))
      };
  }
  private static OrderState ConvertToEnum(string state)
  {
       return state.ToUpperInvariant() switch
       {
           "U" => OrderState.Unpaid,
           "P" => OrderState.Paid,
            ....
           _ => throw new InvalidEnumArgumentException(...)
       };
  }
}

```
Translate to SQL as
```SQL
CAST([p0].[State] ^ N''S'' AS bit) AS [Paid]
```

EF Core version: 9.0.0
Database provider:  Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 8.0
Operating system: Windows 11 10.0.22631
IDE: Visual Studio 2022 17.11.6

**Comment by ranma42** [comment]
I will try and tackle this next weekend (aka it's unlikely I will have a PR ready before next week).

(the same issue is likely also affecting the `~` operator)

**Comment by ChrisJollyAU** [comment]
@roji @ranma42 integer data types are fine as well https://learn.microsoft.com/en-us/sql/t-sql/language-elements/bitwise-exclusive-or-transact-sql?view=sql-server-ver16

**Comment by ranma42** [comment]
@ChrisJollyAU they are fine in that they support the operators (^ and ~), but they might not match the expected Boolean semantics.
For example one could have mapped `true` to `2` and `false` as `4`, in which case `a ^ b` would result in a value that is neither `true/2` nor `false/4` (same goes for `~`).

**Comment by JamesHill3** [comment]
Is there a work around or a setting that will change this default behavior? All of the enums that we use are stored as strings in the DB for readability. Any query we run now in the application that filters by an enum fails.

**Comment by equipatuequipo** [comment]
I had been waiting for version 9.0.1 to be released to implement this change in my project. Today, when testing it, I realized it is only partially fixed. For some reason, when the return value of the ternary operator is a primitive, everything works fine, but if it returns an object, it still throws the same error as before:

This works:
 ```
await _dbContext
    .Machine
    .Select(x => new
    {
        x.Id,
        Value = x.Type == MachineTypes.MyMachineType ? 1 : 2
    }).ToArrayAsync();
```
This doesn't:
```
await _dbContext
    .Machine
    .Select(x => new
    {
        x.Id,
        Value = x.Type == MachineTypes.MyMachineType ? new { } :  null
    }).ToArrayAsync();
```

x.Type and MachineTypes.MyMachineType are enums.

Could you reopen the issue?  

Thank you very much.

**Comment by ranma42** [comment]
I am unable to reproduce the error on 9.0.1. The full code I am running is:

```csharp
// @nuget: Microsoft.EntityFrameworkCore.SqlServer -Version 9.0.1

using System;
using System.Data;
using System.Linq;
using Microsoft.EntityFrameworkCore;

using var db = new BloggingContext();

var queryInt = db
   .Machine
   .Select(x => new
   {
       x.Id,
       Value = x.Type == MachineTypes.MyMachineType ? 1 : 2
   }).ToQueryString();
Console.WriteLine(queryInt);

var queryObj = db
   .Machine
   .Select(x => new
   {
       x.Id,
       Value = x.Type == MachineTypes.MyMachineType ? new { } :  null
   }).ToQueryString();

Console.WriteLine(queryObj);

public class BloggingContext : DbContext
{
    public DbSet<Machine> Machine { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder options)
        => options
            .LogTo(Console.WriteLine, Microsoft.Extensions.Logging.LogLevel.Information)
            .UseSqlServer();

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
	    modelBuilder.Entity<Machine>().Property(x => x.Type).HasConversion<string>();

		modelBuilder.Entity<Machine>().HasData(new Machine { Id = 1, Type = MachineTypes.MyMachineType });
		modelBuilder.Entity<Machine>().HasData(new Machine { Id = 2, Type = MachineTypes.YourMachineType });
    }
}

public enum MachineTypes
{
	MyMachineType,
	YourMachineType,
}

public class Machine
{
    public int Id { get; set; }
	public MachineTypes Type { get; set; }
}
```

I am running it on https://dotnetfiddle.net/ and with 9.0.1 or 8.0.0 I get:
```
SELECT [m].[Id], CASE
    WHEN [m].[Type] = N'MyMachineType' THEN 1
    ELSE 2
END AS [Value]
FROM [Machine] AS [m]
SELECT [m].[Id], CASE
    WHEN [m].[Type] = N'MyMachineType' THEN CAST(1 AS bit)
    ELSE CAST(0 AS bit)
END
FROM [Machine] AS [m]
```

Conversely on 9.0.0 I get:
```
SELECT [m].[Id], CASE
    WHEN [m].[Type] = N'MyMachineType' THEN 1
    ELSE 2
END AS [Value]
FROM [Machine] AS [m]
SELECT [m].[Id], ~CAST([m].[Type] ^ N'MyMachineType' AS bit)
FROM [Machine] AS [m]
```
i.e. 9.0.0 introduced a regression from 8.x, but AFAICT that has been fixed in 9.0.1

@equipatuequipo can you help me get a reproducer?

**Comment by equipatuequipo** [comment]
Thank you very much @ranma42  for your interest. After conducting several tests, I realized that my dependency project, where the ORM is located, did not have the NuGet package updated to version 9.0.1; it was only updated in the one I was currently working on. Upon compiling, it was using version 9.0.0 of the library. After updating everything, it now works correctly. I apologize for the confusion.

Thank you.  
Best regards.
